net-im/libcommuni: New package

ConiKost opened the pull request:
net-im/libcommuni: New package

Closes: https://bugs.gentoo.org/652462

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,49 @@
+
+# Compiling with USE="examples" or USE="tests" is currently broken with dev-libs/icu
+REQUIRED_USE="^^ ( icu uchardet )
+		examples? ( uchardet )
```
Single tab indent?

ConiKost replied:
Fixed.

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,49 @@
+
+RDEPEND="dev-qt/qtcore:5
+	dev-qt/qtdeclarative:5
+	icu? ( dev-libs/icu )
```
`:=`, this thing breaks ABI every other week ;-).

ConiKost replied:
If I understand this correctly, ```:=``` means, when the lib changed (e.g. new slot version), an automatic rebuild of this package will be requested by portage?

mgorny replied:
Yes, if slot or subslot changes.

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,49 @@
+	# Option 'no_rpath' is needed for fixing the QA Notice
+	eqmake5 libcommuni.pro \
+		-config no_rpath
+		-config $(usex benchmarks 'benchmarks' 'no_benchmarks') \
```
Strictly speaking, you could do `$(usex ... '' 'no_')benchmarks` etc. but up to you.

ConiKost replied:
Thanks, souds good to me.

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,49 @@
+	eqmake5 libcommuni.pro \
+		-config no_rpath
+		-config $(usex benchmarks 'benchmarks' 'no_benchmarks') \
+		-config $(usex debug 'debug' 'release') \
```
What does this do? Because it looks suspicious as to another default CFLAGS setter.

ConiKost replied:
If I do understand the source code correctly, it only enables some debug messages for output, when using this lib. I didn't saw any new set CFLAGS.

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,49 @@
+}
+
+src_install() {
+	# INSTALL_ROOT instead of DESTDIR is needed
```
Yep, that's normal for qmake.

ConiKost replied:
It there any other command for running qmake this way? Or do I have to specify it, as it is now done?

mgorny commented on net-im/libcommuni/metadata.xml:
```diff
@@ -0,0 +1,27 @@
+		<flag name="benchmarks">Install optional benchmarks.</flag>
+		<flag name="debug">Build as a debug release.</flag>
+		<flag name="examples">Install optional examples.</flag>
+		<flag name="icu">Use dev-libs/icu instead of app-i18n/uchardet.</flag>
```
'instead of' does not really fit here since you still need to enable the other flag.

ConiKost replied:
I will leave it here. Instead, I will drop uchardet as a use flag. Uchardet is anyway the default, used by libcommuni. So anyone, who sets now icu, will be use this instead of uchardet.

mgorny replied:
Well, I would actually suggest doing the opposite. Given that uchardet is preferred, `IUSE=+uchardet` is cleaner in the intent, while `IUSE=icu` may be accidentally enabled via over-eager global USE flags.

ConiKost replied:
Hm, yeah, makes more sense. I changed the logic now.

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,49 @@
+		-config $(usex qml 'install_qml' 'no_install_qml') \
+		-config $(usex tests 'tests' 'no_tests') \
+		-config $(usex uchardet 'uchardet' 'no_uchardet') \
+		$(usex static '-config static' '')
```
Please remove this. We don't do static.

ConiKost replied:
Okay, fine for me. May I ask, why not provide such an option for an user?

mgorny replied:
Static builds are a security nightmare and we do not provide any means of protecting users from vulnerabilities due to linking statically to old versions. Not to mention they can break randomly on upgrades, and there is no purpose in using static linking for major UI thingies.

mgorny commented on net-im/libcommuni/metadata.xml:
```diff
@@ -0,0 +1,27 @@
+		<flag name="icu">Use dev-libs/icu instead of app-i18n/uchardet.</flag>
+		<flag name="qml">Install optional qml files.</flag>
+		<flag name="static">Build as a static release.</flag>
+		<flag name="tests">Install optional tests.</flag>
```
Why would you install tests instead of running them?

ConiKost replied:
Right, small typo :(

ConiKost commented:
Pushed a new version. I also changed USE="tests" to USE="test", so this should better fit the standard.

ConiKost commented:
1. Fixed by removing "silent" from the *.pri files.
2) Fixed by modifying the *.pri file for uchardet, to use the system library.
3) Fixed. But I am confused. How does this work? I don't understand this.
4) Fixed by dropping it.
5) Fixed by dropping it.
6) Looking into the source, I think, we can drop debug. The debug messages seems not to depend on debug itself. I will remove it.

mgorny commented on net-im/libcommuni/files/uchardet.pri:
```diff
@@ -0,0 +1,17 @@
+isEmpty(PKGCONFIG) {
+	error("UChardet support has been enabled, but the UChardet installation has not been found. Did you emerged app-i18n/uchardet?")
+} else {
+	LIBS += -luchardet
```
Aren't you using pkgconfig already? Or am I misguessing the meaning of the above.

ConiKost replied:
To be honest, I don't know. I just took `/src/3rdparty/icu/icu.pri` as a template and done it the same way.

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,66 @@
+	default
+
+	# Don't use bundled uchardet, use instead app-i18n/uchardet
+	if use uchardet; then
```
I think you can do this unconditionally. Conditional patching is bad since it's easy to miss mistakes when using the other configuration.

ConiKost replied:
If that is okay for you, I would like to do this. I did only conditionally, as I thought, you would find this better ;-)

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,66 @@
+	# Don't compile silent
+	local pri_files=( "src/imports/imports.pri" "src/module_build.pri" )
+	for pri_file in "${pri_files[@]}"; do
+		sed -i -e 's/!verbose:CONFIG += silent//' ${pri_file} || die
```
Can't you just enable that `verbose` thing?

ConiKost replied:
That whould be too easy :P But yes, it apparently works this way fine. Will change it.

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,66 @@
+}
+
+src_configure() {
+	# Option 'no_benchmarks' disables compile run-time benchmarks, as they don't make any sense
```
This is where you realize that using an array would be better as you can then mix comments with options (and also frees you from backslash hell).

ConiKost replied:
I guess, you won't like me anymore ;-), when I say, I don't like mix comments with options. But to make Gentoo more happy, I changed it to an array, as I agree on the backslash hell :D

ConiKost commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,57 @@
+
+src_configure() {
+	local myqmakeargs=( "libcommuni.pro"
+		"-config no_benchmarks" # Diables compile run-time benchmarks, as they don't make any sense
```
Ah okay. Changed it :)

ConiKost commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,57 @@
+		"-config no_examples" # Disables examples, as no new files are installed
+		"-config no_rpath" # Is needed for fixing the QA Notice: The following files contain insecure RUNPATHs
+		"-config release" # Compile libcommuni always in Release mode, as Debug seems to do nothing
+		"-config verbose" # Don't silent all compile messages
```
Thanks, fixed.

ConiKost commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,57 @@
+		"-config $(usex uchardet 'no_' '')icu"
+		"-config $(usex uchardet '' 'no_')uchardet" )
+
+	eqmake5 ${myqmakeargs[@]}
```
Thanks, fixed.

ConiKost commented on net-im/libcommuni/metadata.xml:
```diff
@@ -0,0 +1,22 @@
+	</longdescription>
+	<use>
+		<flag name="qml">Install optional qml files.</flag>
+		<flag name="test">Run optional tests.</flag>
```
Thanks, dropped.

ConiKost commented on net-im/libcommuni/metadata.xml:
```diff
@@ -0,0 +1,22 @@
+	<use>
+		<flag name="qml">Install optional qml files.</flag>
+		<flag name="test">Run optional tests.</flag>
+		<flag name="uchardet">Use <pkg>app-i18n/uchardet</pkg> instead of <pkg>dev-libs/icu</pkg>.</flag>
```
Thanks, fixed.

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,62 @@
+src_configure() {
+	local myqmakeargs=( "libcommuni.pro"
+		# Disables compile run-time benchmarks, as they don't make any sense
+		"-config no_benchmarks"
```
Does having this quoted into single parameter actually work?

mgorny commented:
Good news is that I have no further comments to the ebuild. Bad news is that the tests have started failing for some reason:

```
FAIL!  : tst_IrcMessage::testDecoder(Shift-JIS) Compared values are not the same
   Actual   (actual)  : "\u2022\u00B6\u017D\u0161\u0083R\u0081[\u0083h\u2022\\\u0083V\u0083t\u0083g"
   Expected (expected): "\u6587\u5B57\u30B3\u30FC\u30C9\u8868\u30B7\u30D5\u30C8"
   Loc: [tst_ircmessage.cpp(276)]
```

I thought that stopping to use bundled uchardet was the cause at first but after reenabling it, it fails the same. Could you try investigating the problem? Alternatively, I can commit it as-is if you promise me to take care of it later.

mgorny commented:
[net-im:libcommuni-3.5.0:20180502-152535.log](https://github.com/gentoo/gentoo/files/1967923/net-im.libcommuni-3.5.0.20180502-152535.log)

ConiKost commented:
As we have spoken, the problem happens, when libcommuni is already installed with USE="-uchardat" and new emerge is with USE="uchardet", or the other way around. This is due target_wrapper.sh, which include LD_LIBRARY_PATH, to include libs from /usr/lib64, instead of lib from the compile path. As the "target_wrapper.sh" is generated by qmake not libcommuni itself, I don't really know, what to do. As of now, I just wrote an sed oneliner, to fix those LD_LIBRARY_PATH, which makes the test now pass. I am aware, that this is an ugly hack, but I think, for now, its better than nothing..

mgorny commented on net-im/libcommuni/libcommuni-3.5.0.ebuild:
```diff
@@ -0,0 +1,73 @@
+	# because it uses the system installed lib instead of the new compiled lib.
+	# The test will fail, when libcommuni is emerged with USE="uchardet" and
+	# libcommuni is already installed with USE="-uchardat", or the other way around.
+	find "${S}" -type f -name 'target_wrapper.sh' -exec sed -i -e "s/\/usr\/lib64\${LD_LIBRARY_PATH:+:\$LD_LIBRARY_PATH}/${S//\//\\/}\/lib/" {} \; || die
```
@gentoo/qt, could you advise here? Is there a better way to make qmake behave sanely?

Pesa replied:
Define "behave sanely"... Are you talking about setting an rpath on the test executables? In that case you can use `QMAKE_RPATHDIR`. Other than that, qmake doesn't do anything special on its own AFAIK.

ConiKost replied:
The generated `target_wrapper.sh` includes ` LD_LIBRARY_PATH=/usr/lib64${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}`

mgorny replied:
This sed really looks fragile. Instead, please remove LD_LIBRARY_PATH from that wrapper, and set it in ebuild.

ConiKost replied:
Done.

mgorny replied:
I menat: `d` the whole line from that file, and set `local -x LD_LIBRARY_PATH=...` inside ebuild. Passing paths to sed expressions is a really bad idea.

ConiKost replied:
Okay, changed.

mgorny commented:
Thanks a lot! Looks good now, I'm gonna merge.
